src/App.jsx

```diff
@@ -1,247 +1,309 @@
-              </span>
-
-              <span className="icon is-right">
-                {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
-                <button
-                  data-cy="ClearButton"
-                  type="button"
-                  className="delete"
-                />
-              </span>
+import cn from 'classnames';
+
+import usersFromServer from './api/users';
+import categoriesFromServer from './api/categories';
+import productsFromServer from './api/products';
+
+const products = productsFromServer.map((product) => {
+  const category = categoriesFromServer
+    .find(({ id }) => id === product.categoryId);
+
+  const user = usersFromServer.find(({ id }) => id === category.ownerId);
+
+  return {
+    id: product.id,
+    productName: product.name,
+    title: category.title,
+    icon: category.icon,
+    sex: user.sex,
+    userName: user.name,
+  };
+});
+
+const filterProduct = (userFilter, inputState, categoriesList) => {
+  if (!userFilter && !inputState && !categoriesList.length) {
+    return products;
+  }
+
+  let prods = [...products];
+
+  if (userFilter) {
+    prods = prods.filter(({ userName }) => userName === userFilter);
+  }
+
+  const newInputValue = inputState.trim().toLowerCase();
+
+  if (inputState) {
+    prods = prods.filter(({ productName }) => productName
+      .trim().toLowerCase().includes(newInputValue));
+  }
+
+  if (categoriesList.length) {
+    prods = prods.filter(({ title }) => categoriesList.includes(title));
+  }
+
+  return prods;
+};
+
+const sortProducts = (prevProds, sortState, sortCoefficient) => {
+  if (!sortCoefficient) {
+    return prevProds;
+  }
+
+  const copyProds = [...prevProds];
+
+  switch (sortState) {
+    case 'ID': return copyProds.sort((a, b) => (a.id - b.id) * sortCoefficient);
+    case 'Product': return copyProds
+      .sort((a, b) => (a.productName
+        .localeCompare(b.productName)) * sortCoefficient);
+    case 'Category': return copyProds
+      .sort((a, b) => (a.title
+        .localeCompare(b.title)) * sortCoefficient);
+    case 'User': return copyProds
+      .sort((a, b) => (a.userName
+        .localeCompare(b.userName)) * sortCoefficient);
+
+    default: throw new Error('Error');
+  }
+};
+
+const getSortValue = (currentCoeff) => {
+  if (currentCoeff === 0) {
+    return 1;
+  }
+
+  if (currentCoeff === 1) {
+    return -1;
+  }
+
+  return 0;
+};
+
+const sortFields = ['ID', 'Product', 'Category', 'User'];
+
+export const App = () => {
+  const [userFilter, setUserFilter] = useState('');
+  const [inputState, setInputState] = useState('');
+  const [categoriesList, setCategoriesList] = useState([]);
+  const [sortState, setSortState] = useState('');
+  const [sortCoefficient, setSortCoefficient] = useState(0);
+
+  let prods = filterProduct(userFilter, inputState, categoriesList);
+
+  prods = sortProducts(prods, sortState, sortCoefficient);
```

Comment:
you can create one general function for items preparing
and inside those function call the other needed ones